lib/docker-toolbelt.coffee

```diff
@@ -42,45 +49,106 @@ createChainIdFromParent = (parent, dgsts) ->
 
 	return createChainIdFromParent(dgst, dgsts[1..])
 
+getDiffIds = Promise.method (dkroot, driver, imageId) ->
+	[ hashType, hash ] = imageId.split(':')
+	fs.readFileAsync(path.join(dkroot, "image/#{driver}/imagedb/content", hashType, hash))
+	.then(JSON.parse)
+	.get('rootfs').get('diff_ids')
+
+getCacheId = Promise.method (dkroot, driver, layerId) ->
+	[ hashType, hash ] = layerId.split(':')
+	cacheIdPath = path.join(dkroot, "image/#{driver}/layerdb", hashType, hash, 'cache-id')
+	# Resolves with 'rootId'
+	fs.readFileAsync(cacheIdPath, encoding: 'utf8')
+
 # Gets an string `image` as input and returns a promise that
 # resolves to the absolute path of the root directory for that image
+#
+# Note: in aufs, the path corresponds to the directory for only
+# the specific layer's fs.
 Docker::imageRootDir = (image) ->
-	Promise.all [
+	Promise.join(
 		@infoAsync()
 		@versionAsync().get('Version')
 		@getImage(image).inspectAsync()
-	]
-	.spread (dockerInfo, dockerVersion, imageInfo) ->
-		dkroot = dockerInfo.DockerRootDir
-
-		imageId = imageInfo.Id
-
-		Promise.try ->
-			if semver.lt(dockerVersion, '1.10.0')
-				return imageId
-
-			[ hashType, hash ] = imageId.split(':')
-
-			fs.readFileAsync(path.join(dkroot, "image/#{dockerInfo.Driver}/imagedb/content", hashType, hash))
-			.then(JSON.parse)
-			.then (metadata) ->
-				layerId = createChainId(metadata.rootfs.diff_ids)
-				[ hashType, hash ] = layerId.split(':')
-
-				cacheIdPath = path.join(dkroot, "image/#{dockerInfo.Driver}/layerdb", hashType, hash, 'cache-id')
-
-				# Resolves with 'rootId'
-				fs.readFileAsync(cacheIdPath, encoding: 'utf8')
-		.then (destId) ->
-			switch dockerInfo.Driver
-				when 'btrfs'
-					path.join(dkroot, 'btrfs/subvolumes', destId)
-				when 'overlay'
-					imageInfo.GraphDriver.Data.RootDir
-				when 'vfs'
-					path.join(dkroot, 'vfs/dir', destId)
-				else
-					throw new Error("Unsupported driver: #{dockerInfo.Driver}/")
+		(dockerInfo, dockerVersion, imageInfo) ->
+			dkroot = dockerInfo.DockerRootDir
+
+			imageId = imageInfo.Id
+
+			Promise.try ->
```

Comment:
There's no need for this `Promise.try`, since we're already inside a promise chain you can just return directly in the callback